Analytical_Engine/Query/IsRectangle.cs

```diff
@@ -0,0 +1,113 @@
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Analytical.Elements;
+using BH.oM.Geometry;
+using BH.oM.Reflection.Attributes;
+using BH.Engine.Geometry;
+using BH.Engine.Reflection;
+using BH.Engine.Spatial;
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.ComponentModel;
+
+namespace BH.Engine.Analytical
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Determines whether a panel is a rectangle")]
+        [Input("panel", "The Panel to check if it is a rectangle")]
+        [Output("bool", "True for rectangular panels or false for non-rectangular panels")]
+        public static bool IsRectangle<TEdge, TOpening>(this IPanel<TEdge, TOpening> panel)
+            where TEdge : IEdge
+            where TOpening : IOpening<TEdge>
+        {
+            //Check if there are any openings
+            if (panel.Openings.Count != 0)
+                return false;
+
+            List<ICurve> curves = panel.ExternalEdges.SelectMany(x => x.Curve.ISubParts()).ToList();
+
+            List<PolyCurve> polycurves = Engine.Geometry.Compute.IJoin(curves);
+
+            //Check there is a single continuous curve defining the Panel
+            if (polycurves.Count != 1)
+                return false;
+
+            PolyCurve polycurve = polycurves.First();
+
+            //Check that all subparts of the curve are linear
+            if (polycurve.SubParts().Any(x => !x.IIsLinear()))
+                return false;
+
+            //Group curves by direction vector to obtain discontinuity points
+            List<Point> points = new List<Point>();
+            var groupedCurves = curves.GroupBy(x => x.IStartDir());
+            foreach (var groupedCurve in groupedCurves)
+            {
+                ICurve jointCurve = Engine.Geometry.Compute.IJoin(groupedCurve.Select(x => x).ToList()).First();
+                points.Add(jointCurve.IStartPoint());
+            }
+
+            //Check there are four discontinuity points present (start and end same point)
+            if (points.Count != 4)
+                return false;
```

Comment:
If there are 4 points, and the first and last point at the same, is that not now a triangle rather than a rectangle?

Reply:
Think this will be checked below, this is just a first step checking if it is worth to continue.

Reply:
But we return false if the points are not equal to 4, when something being equal to 4 is a triangle, so what would be the point of the check below? Maybe I'm missing something?

Reply:
If you have 3 or less points, you can be sure it is not a rectangle. If it is 5 or more points, you can be sure it is not a rectangle. -> return false.

If you have 4 points, it might be a rectangle, do more check to determine if this is the case.

That is at least the logic I see behind this.

Reply:
And now I am reading what you are reading for the comment above the if-statement. Agree, if the comment is right, the if-statement should say `!= 5`. If not, the comment should be fixed.

Reply:
I willl ammend the comment.

It would be `!=5` if we used `.IDiscontinuityPoints()` to get the points, but for a `PolyCurve` this returns intermediate points that exist on the same line (see test script where square/rectangle is constructed from >4 lines).

So I've had to group the curves by direction, joined them and taken the start point of each edge resulting in four points being returned.